lib/content_security_policy/extension.rb

```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+class ContentSecurityPolicy
+  module Extension
+    extend self
+
+    def site_setting_extension
+      { script_src: SiteSetting.content_security_policy_script_src.split('|') }
+    end
+
+    def plugin_extensions
+      DiscoursePluginRegistry.csp_extensions
+    end
+
+    THEME_SETTING = 'extend_content_security_policy'
+
+    def theme_extensions
+      cache['theme_extensions'] ||= find_theme_extensions
+    end
+
+    def clear_theme_extensions_cache!
+      cache['theme_extensions'] = nil
+    end
+
+    private
+
+    def cache
+      @cache ||= DistributedCache.new('csp_extensions')
+    end
+
+    def find_theme_extensions
+      extensions = []
+
+      Theme.find_each do |theme|
```

Comment:
I initially wanted to optimize this into something like 

```rb
Theme.join(:theme_settings).where(theme_settings: { name: THEME_SETTING })
```

But this won't work because we do not create a `theme_setting` entry for empty/non-overwritten values.

Reply:
I was also thinking if I should cache the final combined CSP instead, but that means a lot more cache invalidation cases.

Reply:
I would not stress about this, it is cached

Reply:
regarding caching the entire CSP... its probably ok not to, but you can measure the impact... if less than say 0.5ms its probably not worth it.

Reply:
Measured and it's consistently ~0.1ms when theme extensions are cached. Going to leave this as is 💪 Thanks